### DISCUSSION TITLE
How to find the group position of a widget without an explicit reference to it's parent

### DISCUSSION THREAD

**Discussion by wyhinton** [fltk-rs/fltk-rs#749] [Q&A]
Say we have a custom widget ```MyButton```, that we'll be adding to a ```Pack```. Is there anyway to get the index of ```MyButton``` without passing the parent ```Pack``` through the ```MyButton``` constructor? 

I.e being able to use ```let index = button.parent().unwrap().find(&button);```?

The problem is that ```parent().unwrap()``` returns a ```Box<dyn GroupExt>```, and we can't use ```find()``` on a trait object. But there is no way to up-cast our ```Box<dyn GroupExt>``` to a ```Group``` like we can up-cast a ```Box<dyn WidgetExt>``` to a ```Widget``` via ```as_widget_ptr()```.  Is their a work around for this?

```rust
use fltk::{app, app::*, group::Pack, group::*, self, button::*, draw, enums::*, frame::*, prelude::*, widget::*, window::*};
pub struct MyButton{
  but: button
}

impl MyButton{
    pub fn new(x: i32, y: i32, w: i32, h: i32)->Self{
        //if we pass the parent through the constructor then the group is not boxed, but this makes our constructor bulkier
        let mut button= Button::new(x,y,w,h,None);
        //find the position of the button within it's parent group and set this as it's label
        let index = button.parent().unwrap().find(&button);
        button.set_label(&index.to_string());
        MyButton{
            but: button
        }
    }
}

fn main() {
    let app = App::default();
    let mut win = Window::new(200, 200, 1000, 1000, "Template");
    let container= Pack::new(0,0,200,200,"mypack");    
    let my_but1 = MyButton::new(0,0,100,20);
    let my_but2 = MyButton::new(0,0,100,20);
    container.end();
    win.end();
    win.show();
    app.run().unwrap();
}
```
Results in error:
```
error: the `find` method cannot be invoked on a trait object
```

**Comment by MoAlyousef** [comment]
That's a limitation in Rust which disallows generic methods on trait objects.
I can add an `into_group()` method which allows such conversions:
```rust
let index = unsafe { button.parent().unwrap().into_group().find(&button) };
```
It has to be unsafe, since it allows creating a concrete type from an erased type which might not have been created by fltk-rs. That means that methods such as handle and draw can't be overriden.

**Reply by wyhinton** [reply]
Yes, I think adding an unsafe ```into_group()``` makes sense if we also have an unsafe ```into_widget()```
